src/main/java/net/spy/memcached/tls/TLSConnectionManager.java

```diff
@@ -0,0 +1,240 @@
+            }
+            closeSslEngine();
+        }
+        sslEngine = sslContext.createSSLEngine();
+        sslEngine.setUseClientMode(true);
+    }
+
+    private void initBuffers(SSLSession session) {
+        // allocateDirect() to keep all bytes contiguous in memory
+        appOutBuffer = ByteBuffer.allocateDirect(session.getApplicationBufferSize());
+        appInBuffer = ByteBuffer.allocateDirect(session.getApplicationBufferSize());
+        networkOutBuffer = ByteBuffer.allocateDirect(session.getPacketBufferSize());
+        networkInBuffer = ByteBuffer.allocateDirect(session.getPacketBufferSize());
+    }
+
+    public boolean doHandshake(SocketChannel socketChannel) throws IOException {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("%s - Beginning handshake.", socketChannel.getRemoteAddress());
+        }
+
+        try {
+            initSSLEngine();
+            sslEngine.beginHandshake();
+            currentSession = sslEngine.getSession();
+            initBuffers(currentSession);
+
+            HandshakeStatus handshakeStatus = sslEngine.getHandshakeStatus();
+            if (LOG.isDebugEnabled()) {
+                LOG.info("%s - Handshake status: %s", socketChannel.getRemoteAddress(), handshakeStatus.name());
+            }
+            while (!Thread.currentThread().isInterrupted()
+                    && handshakeStatus != HandshakeStatus.FINISHED
+                    && handshakeStatus != HandshakeStatus.NOT_HANDSHAKING) {
+
+                switch (handshakeStatus) {
+                    case NEED_TASK:
+                        // we need to finish these tasks for the handshake to continue
+                        List<Future<?>> tasks = new ArrayList<>();
+                        Runnable currentTask = sslEngine.getDelegatedTask();
+                        while(currentTask != null) {
+                            tasks.add(executor.submit(currentTask));
+                            currentTask = sslEngine.getDelegatedTask();
+                        }
+                        for (Future<?> task : tasks) {
+                            try {
+                                task.get();
+                            } catch (ExecutionException e) {
+                                // TODO make this more clear
+                                throw new RuntimeException(e);
+                            }
+                        }
+                        break;
+                    case NEED_WRAP:
+                        // We're about to send something to the server, but it needs to be wrapped first
+                        networkOutBuffer.clear();
+                        SSLEngineResult wrapResult = sslEngine.wrap(appOutBuffer, networkOutBuffer);
+                        handleHandshakeWrapResult(socketChannel, wrapResult);
+                        break;
+                    case NEED_UNWRAP:
+                        // We need to receive a message from the server, but it needs to be unwrapped first
```

Comment:
I believe we need networkInBuffer.clear() here like we have for NEED_WRAP

Reply:
When we receive data from the network, if it's not the entire message, we need to call `.compact()` and let the loop read more from the network. If we `.clear()` here, we'll blow away data that we're still reading and get stuck in this loop forever.

Reply:
hmm should it be like
```
 if (!networkInBuffer.hasRemaining()) {
        networkInBuffer.clear();
    } else {
        networkInBuffer.compact();
    }
```

Reply:
For some reason, adding that code causes the handshake to fail with
```
Error parsing certificate message:no sufficient data
```